File: src/main/java/com/topface/topface/ui/bonus/view/BonusFragment.java
package com.topface.topface.ui.bonus.view;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.topface.framework.utils.Debug;
import com.topface.topface.App;
import com.topface.topface.R;
import com.topface.topface.databinding.FragmentBonusBinding;
import com.topface.topface.mvp.IPresenterFactory;
import com.topface.topface.mvp.PresenterCache;
import com.topface.topface.statistics.FlurryOpenEvent;
import com.topface.topface.ui.adapters.ItemEventListener;
import com.topface.topface.ui.bonus.models.IOfferwallBaseModel;
import com.topface.topface.ui.bonus.presenter.BonusPresenter;
import com.topface.topface.ui.bonus.presenter.IBonusPresenter;
import com.topface.topface.ui.bonus.viewModel.BonusFragmentViewModel;
import com.topface.topface.ui.fragments.BaseFragment;
import com.topface.topface.ui.views.toolbar.utils.ToolbarManager;
import com.topface.topface.ui.views.toolbar.utils.ToolbarSettingsData;
import com.topface.topface.utils.Utils;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

@FlurryOpenEvent(name = BonusFragment.PAGE_NAME)
public class BonusFragment extends BaseFragment implements IBonusView {
    public static final String NEED_SHOW_TITLE = "need_show_title";
    public static final String PAGE_NAME = "bonus";
    public static final String TAG = BonusFragment.class.getSimpleName();

    private FragmentBonusBinding mBinding;
    private IBonusPresenter mPresenter;
    private OfferwallsAdapter mAdapter;
    private PresenterCache mPresenterCache;

    public static BonusFragment newInstance(boolean needShowTitle) {
        BonusFragment fragment = new BonusFragment();
        Bundle arguments = new Bundle();
        arguments.putBoolean(NEED_SHOW_TITLE, needShowTitle);
        fragment.setArguments(arguments);
        return fragment;
    }

    @Override
    protected String getScreenName() {
        return PAGE_NAME;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mPresenterCache = App.getAppComponent().presenterCache();
        mPresenter = mPresenterCache.getPresenter(TAG, new IPresenterFactory<IBonusPresenter>() {
            @NotNull
            @Override
            public IBonusPresenter createPresenter() {
                return new BonusPresenter();
            }
        });
        mBinding = DataBindingUtil.bind(inflater.inflate(R.layout.fragment_bonus, null));
        BonusFragmentViewModel viewModel = new BonusFragmentViewModel();
        mPresenter.setViewModel(viewModel);
        mBinding.setViewModel(viewModel);
        mBinding.rvOfferwalls.setLayoutManager(new LinearLayoutManager(getActivity().getApplicationContext()));
        mBinding.rvOfferwalls.setAdapter(getAdapter());
        mPresenter.bindView(this);
        mPresenter.loadOfferwalls();
        return mBinding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        Bundle arg = getArguments();
        if (arg != null && arg.getBoolean(NEED_SHOW_TITLE)) {
            ToolbarManager.INSTANCE.setToolbarSettings(new ToolbarSettingsData(getString(R.string.general_bonus)));
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.unbindView();
    }

    @Override
    public void showOffers(ArrayList<IOfferwallBaseModel> offers) {
        OfferwallsAdapter adapter = getAdapter();
        adapter.clearData();
        adapter.addData(offers);
    }

    private OfferwallsAdapter getAdapter() {
        if (mAdapter == null) {
            mAdapter = new OfferwallsAdapter();
            mAdapter.setOnItemClickListener(new ItemEventListener.OnRecyclerViewItemClickListener<IOfferwallBaseModel>() {
                @Override
                public void itemClick(View view, int itemPosition, IOfferwallBaseModel data) {
                    Debug.showChunkedLogDebug("Offerwalls", "try open " + data.getOfferwallsType() + " url " + data.getLink());
                    Utils.goToUrl(getActivity(), data.getLink());
                }
            });
        }
        return mAdapter;
    }
}

File: src/main/java/com/topface/topface/ui/bonus/BonusFragment.kt
package com.topface.topface.ui.bonus

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.topface.topface.App
import com.topface.topface.R
import com.topface.topface.databinding.FragmentBonusEnhancedBinding
import com.topface.topface.ui.bonus.components.LoaderComponent
import com.topface.topface.ui.bonus.components.OfferwallButtonComponent
import com.topface.topface.ui.bonus.models.BonusViewModel
import com.topface.topface.ui.fragments.BaseFragment
import com.topface.topface.ui.new_adapter.enhanced.CompositeAdapter
import com.topface.topface.ui.views.toolbar.utils.ToolbarManager
import com.topface.topface.ui.views.toolbar.utils.ToolbarSettingsData
import org.jetbrains.anko.layoutInflater


/**
 * Created by ppavlik on 02.06.17.
 * Фрагмент "разводящего" экрана для оферов
 */
class BonusFragment : BaseFragment() {

    companion object {
        const val PAGE_NAME = "bonus"
        const val NEED_SHOW_TITLE = "need_show_title"
        const val FROM = "from"
        val TAG = BonusFragment::class.java.simpleName

        fun newInstance(isNeedTitle: Boolean, from: String) = BonusFragment().apply {
            arguments = Bundle().apply {
                putBoolean(NEED_SHOW_TITLE, isNeedTitle)
                putString(FROM, from)
            }
        }
    }

    private val mBinding by lazy {
        DataBindingUtil.inflate<FragmentBonusEnhancedBinding>(context.layoutInflater,
                R.layout.fragment_bonus_enhanced, null, false)
    }

    private val mViewModel by lazy {
        BonusViewModel()
    }

    private val mIronSourceManager by lazy {
        App.getAppComponent().ironSourceManager()
    }

    private val mProvider by lazy {
        BonusProvider()
    }

    private val mAdapter: CompositeAdapter by lazy {
        CompositeAdapter(mProvider) { Bundle() }
                .addAdapterComponent(LoaderComponent())
                .addAdapterComponent(OfferwallButtonComponent(arguments.getString(FROM)))
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mIronSourceManager.initSdk(activity)
        initList()
        return mBinding.apply {
            viewModel = mViewModel
        }.root
    }

    override fun onResume() {
        super.onResume()
        if (arguments?.getBoolean(NEED_SHOW_TITLE) ?: false) {
            ToolbarManager.setToolbarSettings(ToolbarSettingsData(getString(R.string.general_bonus)))
        }
    }

    private fun initList() = with(mBinding.list) {
        layoutManager = LinearLayoutManager(context)
        adapter = mAdapter
    }
}